### PR TITLE
BUG: tuple-of-tuples indexing results in NumPy VisibleDeprecationWarning

### LINKED ISSUE #35434: BUG: Shift on a group column when column name is a tuple-of-tuples results in NumPy VisibleDeprecationWarning
- [+] I have checked that this issue has not already been reported.

- [+] I have confirmed this bug exists on the latest version of pandas.

- [+] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
import numpy as np
from datetime import datetime, timedelta

times =             np.arange(datetime(2020, 1, 1), datetime(2020, 2, 1), timedelta(minutes=30)).astype(datetime)
tlen =              len(times)
A, Z =              np.array(['A', 'Z']).view('int32')
symbol_names =      np.random.randint(low=A, high=Z, size=2 * 2, dtype='int32').view(f'U{2}')
times =             np.concatenate([times] * 2)
names =             np.array([y for x in [[s] * tlen for s in symbol_names] for y in x])
first_column =      np.random.randn(len(times))
second_column =     np.random.randn(len(times))

tuple_column =      ('ABC', ('DEF', 444),)
df =                pd.DataFrame({'first_column': first_column, tuple_column: second_column}, index=[times, names])
grp =               df.groupby(level=0)
df[tuple_column] =  grp[[tuple_column]].shift()
```

#### Problem description

Shifting a group that has a column name as tuple-of-tuples gives VisibleDeprecationWarning.
> C:\Python\lib\site-packages\numpy\core\_asarray.py:83: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray

#### Expected Output

No VisibleDeprecationWarning triggered.

#### Output of ``pd.show_versions()``

<details>
commit           : 6302f7b98ad24adda2d5a98fef3956f04f28039d
python           : 3.8.5.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.18362
machine          : AMD64
processor        : Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United Kingdom.1252

pandas           : 1.1.0rc0+8.g6302f7b98
numpy            : 1.19.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.1.1
setuptools       : 49.2.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.2
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : 4.9.1
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.0rc1+439.g7e9530338
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.50.1
</details>

### PR DESCRIPTION
**Pull request by simonjayhawkins**
- [ ] closes #35434
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/common.py [hunk 1]
```diff
@@ -138,9 +138,12 @@ def is_bool_indexer(key: Any) -> bool:
             return True
     elif isinstance(key, list):
         try:
-            arr = np.asarray(key)
+            # https://github.com/pandas-dev/pandas/issues/35434
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", category=np.VisibleDeprecationWarning)
```
Comment (TomAugspurger) [inline pandas/core/common.py:143]: Getting these can be somewhat expensive. Does this slow down `DataFrame.__getitem__` at all?

Comment (simonjayhawkins) [reply pandas/core/common.py:143]: it's not free, but not significant for `ser[[tup]]`
<details>
```
         1187 function calls (1183 primitive calls) in 0.002 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      4/2    0.000    0.000    0.001    0.000 base.py:289(__new__)
      281    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}
       14    0.000    0.000    0.000    0.000 {built-in method numpy.array}
        4    0.000    0.000    0.000    0.000 {pandas._libs.lib.infer_dtype}
       94    0.000    0.000    0.000    0.000 generic.py:10(_check)
        1    0.000    0.000    0.002    0.002 {built-in method builtins.exec}
        5    0.000    0.000    0.000    0.000 common.py:221(asarray_tuplesafe)
        2    0.000    0.000    0.000    0.000 common.py:97(is_bool_indexer)
        2    0.000    0.000    0.000    0.000 {method 'reduce' of 'numpy.ufunc' objects}
        1    0.000    0.000    0.000    0.000 indexing.py:1257(_validate_read_indexer)
        7    0.000    0.000    0.000    0.000 warnings.py:458(__enter__)
        6    0.000    0.000    0.000    0.000 _dtype.py:321(_name_get)
        1    0.000    0.000    0.000    0.000 {pandas._libs.algos.take_1d_int64_int64}
       20    0.000    0.000    0.000    0.000 common.py:1460(is_extension_array_dtype)
        1    0.000    0.000    0.000    0.000 algorithms.py:1586(take_nd)
        1    0.000    0.000    0.000    0.000 {method 'get_indexer' of 'pandas._libs.index.IndexEngine' objects}
       23    0.000    0.000    0.000    0.000 base.py:256(is_dtype)
        1    0.000    0.000    0.002    0.002 series.py:910(_get_with)
        1    0.000    0.000    0.000    0.000 base.py:2951(get_indexer)
      144    0.000    0.000    0.000    0.000 {built-in method builtins.getattr}
       20    0.000    0.000    0.000    0.000 base.py:413(find)
        7    0.000    0.000    0.000    0.000 warnings.py:181(_add_filter)
        1    0.000    0.000    0.000    0.000 generic.py:4493(_reindex_with_indexers)
        2    0.000    0.000    0.000    0.000 cast.py:441(maybe_promote)
        1    0.000    0.000    0.002    0.002 series.py:868(__getitem__)
        1    0.000    0.000    0.002    0.002 indexing.py:1078(_getitem_axis)
       21    0.000    0.000    0.000    0.000 common.py:1600(_is_dtype_type)
        2    0.000    0.000    0.000    0.000 base.py:5718(_maybe_cast_data_without_dtype)
        7    0.000    0.000    0.000    0.000 dtypes.py:1113(is_dtype)
        1    0.000    0.000    0.000    0.000 managers.py:1267(_slice_take_blocks_ax0)
        2    0.000    0.000    0.000    0.000 cast.py:1559(construct_1d_object_array_from_listlike)
        3    0.000    0.000    0.000    0.000 {built-in method numpy.empty}
        3    0.000    0.000    0.000    0.000 generic.py:377(_get_axis)
        1    0.000    0.000    0.001    0.001 indexing.py:1208(_get_listlike_indexer)
        1    0.000    0.000    0.000    0.000 blocks.py:1233(take_nd)
       76    0.000    0.000    0.000    0.000 {built-in method builtins.issubclass}
        9    0.000    0.000    0.000    0.000 common.py:1565(_get_dtype)
        6    0.000    0.000    0.000    0.000 _dtype.py:24(_kind_name)
        9    0.000    0.000    0.000    0.000 common.py:530(is_categorical_dtype)
        7    0.000    0.000    0.000    0.000 dtypes.py:901(is_dtype)
        9    0.000    0.000    0.000    0.000 common.py:492(is_interval_dtype)
        1    0.000    0.000    0.000    0.000 algorithms.py:1457(_get_take_nd_function)
        7    0.000    0.000    0.000    0.000 warnings.py:165(simplefilter)
```
</details>

what's the best approach?

**Review by jreback — CHANGES_REQUESTED** [review]
-1 on catching the warming

this needs either an actual fix in index construction

**Comment by simonjayhawkins** [comment]
> this needs either an actual fix in index construction

to prevent nested tuples in the index?

**Comment by simonjayhawkins** [comment]
I'll close this for now to clear the queue, I'll revisit once on top of 1.1.0 regressions. There could be lots to do here to avoid 7 repeated calls to np.asarray and maintain state for the indexer checks.

**Pull request closed without merging**